src/mgr/DaemonServer.cc

```diff
@@ -1395,7 +1398,36 @@ bool DaemonServer::_handle_command(
 	      stored_pgs.insert(osd);
 	      continue;
 	    }
-	  }
+            if (osdmap.is_down(osd) &&
+                osdmap.is_out(osd) &&
+                any_degraded_pgs) {
+              // sadly we totally lost pgmap-related track for down and out osds
+              // so if there is any degraded pgs, we need to fake-mark osd up&in
+              // and continue to check if the degradedness is actually caused by
+              // this osd
+              OSDMap tmpmap;
+              tmpmap.deepish_copy_from(osdmap);
+              tmpmap.set_state(osd, CEPH_OSD_EXISTS | CEPH_OSD_UP);
+              auto old_weight = osdmap.get_xinfo(osd).old_weight;
+              auto new_weight = old_weight ? old_weight : CEPH_OSD_IN;
+              tmpmap.set_weight(osd, new_weight);
+              for (auto &i : pg_map.pg_stat) {
+                auto& pg = i.first;
+                auto state = i.second.state;
+                if (state & (PG_STATE_DEGRADED | PG_STATE_UNDERSIZED)) {
+                  vector<int> up;
+                  int up_primary;
+                  tmpmap.pg_to_raw_up(pg, &up, &up_primary);
+                  if (std::find(up.begin(), up.end(), osd) != up.end()) {
```

Comment:
I'm not sure checking whether this osd *would* store the PG if it were up+in tells us much.

What we really want to know is if there is any possible way that this OSD has data that we may need to recover.  I think the way to do that is to identify any non-active+clean PG, and see whether this OSD has a copy of that PG.  If it does, then we shouldn't destroy it.

Reply:
> I think the way to do that is to identify any non-active+clean PG, and see whether this OSD has a copy of that PG. If it does, then we shouldn't destroy it.

That's the problem. If that osd is currently down& out, we are going to erase any pg or osd related statistics from it. I can't think of a way of doing this without changing the existing stat-counting behaviour much.